pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java

```diff
@@ -707,73 +709,96 @@ private void completeConnect(int clientProtoVersion, String clientVersion, boole
     }
 
     // According to auth result, send newConnected or newAuthChallenge command.
-    private State doAuthentication(AuthData clientData,
-                                   int clientProtocolVersion,
-                                   String clientVersion) throws Exception {
-
+    private CompletableFuture<Void> doAuthenticationAsync(AuthData clientData, int clientProtocolVersion,
+                                                          String clientVersion) {
         // The original auth state can only be set on subsequent auth attempts (and only
         // in presence of a proxy and if the proxy is forwarding the credentials).
         // In this case, the re-validation needs to be done against the original client
-        // credentials.
-        boolean useOriginalAuthState = (originalAuthState != null);
-        AuthenticationState authState =  useOriginalAuthState ? originalAuthState : this.authState;
-        String authRole = useOriginalAuthState ? originalPrincipal : this.authRole;
-        AuthData brokerData = authState.authenticate(clientData);
-
-        if (log.isDebugEnabled()) {
-            log.debug("Authenticate using original auth state : {}, role = {}", useOriginalAuthState, authRole);
-        }
+        // credentials, but we only can new an authentication state, because some authentication data(TLS, SASL)
+        // based on outside service.
```

Comment:
Sorry, there is some confusion. For `originalAuthentication`, we don't call the authentication checks. Because there is some state in the `AuthenticationState`, which depends on outside service.

https://github.com/apache/pulsar/blob/82237d3684fe506bcb6426b3b23f413422e6e4fb/pulsar-broker-auth-sasl/src/main/java/org/apache/pulsar/broker/authentication/SaslAuthenticationState.java#L64-L67

`originalAuthentication` belongs to the user's client by the proxy forwarded, see https://github.com/apache/pulsar/blob/f3608074f537471545926a84df8a1d061b30692c/pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/ProxyClientCnx.java#L68-L70

Reply:
> For `originalAuthentication`, we don't call the authentication checks.

Isn't this a problem though? We aren't really authenticating the `originalAuthData` if we don't call the `authenticate` method and make sure authentication is "complete". In the `ProxyConnectionToBroker` case, we can send back `AuthChallenge In the event that the proxy is forwarding authentication information, we can issue `AuthChallenge` responses. It might not work so easily in the `ProxyLookupRequests` state.

Reply:
For the current design, the `originalAuthData` is credible by the proxy forwarding authentication data. This authentication data was completed in the proxy and then sent this authentication data to the broker, so we don't check the authentication data in the broker.

For `ProxyLookupRequests` state, see https://github.com/apache/pulsar/pull/17831. I forward the `AuthChallenge` command to the user's client by the ProxyClient and Proxy Server.

Reply:
> For the current design, the `originalAuthData` is credible by the proxy forwarding authentication data. This authentication data was completed in the proxy and then sent this authentication data to the broker, so we don't check the authentication data in the broker.

My primary objection is that the setting required to enter this code block is called `isAuthenticateOriginalAuthData`. Note that we are already verifying most auth data in the case that the `newAuthState` method actually authenticates the data. That leaves out cases like SASL which do not authentication on `AuthenticationState` initialization because it is a multi-stage auth. However, I've just realized that forwarding the authentication information will especially not work with SASL because the proxy only forwards the _last_ `AuthData` that it receives:

https://github.com/apache/pulsar/blob/d6fcdb8e1e192e0d9e2fbe6307d273b27ed2930f/pulsar-proxy/src/main/java/org/apache/pulsar/proxy/server/ProxyConnection.java#L426-L440

This seems like a gap in the protocol where multi-staged authentication is not able to be properly processed through the proxy. However, if its a gap, that means it is something that could be improved without breaking anything.

One question is whether it would help to have a distinct stages where we first authenticate the proxy and then authenticate its original client (when configured to do so). That could help to open up the design and to make the protocol a bit clearer.

Reply:
> This seems like a gap in the protocol where multi-staged authentication is not able to be properly processed through the proxy. However, if its a gap, that means it is something that could be improved without breaking anything.

When using the multi-staged authentication, we need to set `authenticateOriginalAuthData=false` in the broker, and then it works fine, the broker just checks the proxy's authentication data.

> One question is whether it would help to have a distinct stages where we first authenticate the proxy and then authenticate its original client (when configured to do so). 

The protocol needs to be changed, distinguishing between client and proxy.


The authentication logic here is a bit confusing, but for the current design we need to give the correct configuration in our documentation.

Reply:
> When using the multi-staged authentication, we need to set `authenticateOriginalAuthData=false` in the broker, and then it works fine, the broker just checks the proxy's authentication data.

I think this supports my statement that there is a gap in the protocol.